dockstore-webservice/src/main/java/io/dockstore/webservice/resources/EntryResource.java

```diff
@@ -438,79 +510,66 @@ public Entry setDiscourseTopic(@ApiParam(hidden = true) @Parameter(hidden = true
-                    LANGUAGE_VERSION_PROCESSOR_PAGE_SIZE, progress.offset);
-                progress.offset += LANGUAGE_VERSION_PROCESSOR_PAGE_SIZE;
-                workflows.forEach(workflow -> {
-                    LanguageHandlerInterface languageHandlerInterface = LanguageHandlerFactory.getInterface(
-                        workflow.getDescriptorType());
-                    workflow.getWorkflowVersions().stream()
-                        .filter(version -> allVersions || version.getVersionMetadata().getDescriptorTypeVersions().isEmpty())
-                        .forEach(version -> {
-                            final String primaryPath = version.getWorkflowPath();
-                            readSourceFilesAndUpdate(languageHandlerInterface, version, primaryPath);
-                        });
-                });
-            });
+    @Path("/updateOpenData")
+    @RolesAllowed("admin")
+    @UnitOfWork
+    @POST
+    @Operation(operationId = "updateOpenData", description = "Update open data", security = @SecurityRequirement(name = JWT_SECURITY_DEFINITION_NAME))
+    @ApiResponse(responseCode = HttpStatus.SC_OK + "", description = "Number of entries processed",
+        content = @Content(mediaType = MediaType.APPLICATION_JSON, schema = @Schema(implementation = Integer.class)))
+    public int updateOpenData(@ApiParam(hidden = true) @Parameter(hidden = true, name = "user") @Auth User user,
+        @Parameter(description = "Whether to process all versions or only versions without open data already set", in = ParameterIn.QUERY, required = false)
+        @QueryParam("allVersions") @DefaultValue("false") boolean allVersions) {
+        if (lambdaUrlChecker == null) {
+            throw new CustomWebApplicationException("The url checker is not configured; this request cannot be processed.", HttpStatus.SC_INTERNAL_SERVER_ERROR);
         }
-        LOG.info("Completed processing {} workflows", progress.processedEntries);
-        return progress.processedEntries;
+        LOG.info("Processing workflows for open data");
+        final int processedEntries = loadAndProcessEntries(getWorkflows, processWorkflowsForOpenData, allVersions);
+        LOG.info("Completed processing {} entries for open data", processedEntries);
+        return processedEntries;
     }
 
-    private int updateLanguageVersionsForLegacyTools(boolean allVersions) {
-        final LanguageVersionProgress progress = new LanguageVersionProgress();
+    /**
+     * Loads and processes entries in batches, committing each batch as it goes along. The batch
+     * size is {@link #PROCESSOR_PAGE_SIZE}
+     * @param loader
+     * @param processor
+     * @param allVersions
+     * @return the number of entries processed
+     */
+    private <T extends Entry> int loadAndProcessEntries(Function<Integer, List<T>> loader,
+        BiFunction<List<T>, Boolean, Void> processor, Boolean allVersions) {
```

Comment:
No need to change anything, but I spent a little time pondering further generalizations to this code.  Rather than just let it evaporate, I'll dump it here...

A more general implementation subsumes the retrieval code into a stream or iterator, and decouples the notion of a "block" of entries from the operation to be performed on an entry.  So, you end up with something like:

```
private <T extends Entry> int processEntries(Iterator<T> source, Consumer<T> operation, int N) {
  while (source.hasNext()) {
      transactionHelper.transaction(() -> {
          int count = 0;
          while (source.hasNext() && count < N) {
              operation.accept(source.next());
              count++;
          }
          LOG.info("processed {} entries", count);
      });
   }
}
```
Input entries can come from anywhere, whether in pages from the db, or from a list we've already loaded.   They're also easy to filter, and we can process them in chunks sized to our choosing.

Reply:
That does sound better, but I am going to leave it as is, as suggested, to get this merged.